Objective-C/CBLCollection.mm

```diff
@@ -244,4 +263,69 @@ - (BOOL) collectionIsValid: (NSError**)error {
     
 }
 
+- (id<CBLListenerToken>) addCollectionChangeListener: (void (^)(CBLCollectionChange*))listener
+                                               queue: (dispatch_queue_t)queue {
+    if (!_colChangeNotifier) {
+        _colChangeNotifier = [CBLChangeNotifier new];
+        _colObs = c4dbobs_createOnCollection(_c4col, colObserverCallback, (__bridge void *)self);
+    }
+    
+    return [_colChangeNotifier addChangeListenerWithQueue: queue listener: listener delegate: self];
+}
+
+static void colObserverCallback(C4CollectionObserver* obs, void* context) {
+    CBLCollection *c = (__bridge CBLCollection *)context;
+    dispatch_async(c.dispatchQueue, ^{
+        [c postCollectionChanged];
+    });
+}
+
+- (void) postCollectionChanged {
+    CBL_LOCK(self) {
+        if (!_colObs || !_c4col)
+            return;
+        
+        const uint32_t kMaxChanges = 100u;
+        C4DatabaseChange changes[kMaxChanges];
+        bool external = false;
+        C4CollectionObservation obs = {};
+        NSMutableArray* docIDs = [NSMutableArray new];
+        do {
+            // Read changes in batches of kMaxChanges:
+            obs = c4dbobs_getChanges(_colObs, changes, kMaxChanges);
+            if (obs.numChanges == 0 || external != obs.external || docIDs.count > 1000) {
+                if(docIDs.count > 0) {
+                    CBLCollectionChange* change = [[CBLCollectionChange alloc] initWithCollection: self
+                                                                                      documentIDs: docIDs
+                                                                                       isExternal: external];
+                    [_colChangeNotifier postChange: change];
+                    docIDs = [NSMutableArray new];
+                }
+            }
+            
+            external = obs.external;
+            for(uint32_t i = 0; i < obs.numChanges; i++) {
+                NSString *docID =slice2string(changes[i].docID);
+                [docIDs addObject: docID];
+            }
+            c4dbobs_releaseChanges(changes, obs.numChanges);
+        } while(obs.numChanges > 0);
+    }
+}
+
+- (void) removeToken: (id)token {
+    CBL_LOCK(self) {
+        if (![self collectionIsValid: nil])
```

Comment:
I think this check might not be needed. Will it cause any problems if remove the check?

Reply:
right, there shouldn't be. we don't work with c4collection here.